### PR TITLE
Add oauth

### PR DESCRIPTION
**Pull request by Sgtpluck**
This PR does the work to get login working, leveraging cloud.gov's UAA server.

See [the cloud.gov docs](https://cloud.gov/docs/management/leveraging-authentication/) for more details.

It also unpins the buildpack version, as the new ruby buildpack is now the default on cloud.gov

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,3 @@
 class ApplicationController < ActionController::Base
+  skip_before_action :verify_authenticity_token, if: proc { :bypass_auth? }
```
Comment (rahearn) [inline app/controllers/application_controller.rb:2]: [issue] I think this is actually always skipping the check. It should probably be either:

`skip_before_action :verify_authenticity_token, if: proc { bypass_auth? }` or

`skip_before_action :verify_authenticity_token, if: :bypass_auth?` or

`skip_forgery_protection if: :bypass_auth?`

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  # skip_before_action :require_user!
```
Comment (rahearn) [inline app/controllers/sessions_controller.rb:2]: [suggestion (nonblocking)] might be worthwhile to clean up this old comment

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,2 +1,14 @@
...
+    @current_user ||= session[:email]
+  end
+
+  def user_signed_in?
+    !!current_user
+  end
+
+  def bypass_auth?
+    Rails.configuration.x.bypass_auth == true
+  end
```
Comment (rahearn) [inline app/helpers/application_helper.rb:13]: [suggestion (nonblocking)] these all feel more like controller methods to me than helper methods, I'd move them up to `ApplicationController` and then expose the one(s) you need in views with `helper_method :method_name`

File: app/models/session.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  def request_params
+    {
+      code: @code,
+      grant_type: "authorization_code",
+      response_type: "token",
+      client_id: ENV["UAA_CLIENT_ID"] || Rails.application.credentials.uaa_client_id,
+      client_secret: ENV["UAA_SECRET"] || Rails.application.credentials.uaa_client_secret,
+      redirect_uri: ENV["REDIRECT_URI"]
+    }
```
Comment (rahearn) [inline app/models/session.rb:48]: [praise] this is much nicer than the earlier string version

**Review by rahearn — CHANGES_REQUESTED** [review]
Nice work! Take or leave the two suggestions, but I think the issue definitely needs to be looked at again.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (Sgtpluck) [reply app/controllers/application_controller.rb:2]: i don't think that's true, not as it's currently written. (i had double checked in my local env to ensure it was kidding the conditional, and taking those suggestions leads to an error `undefined method `bypass_auth?' for #<PagesController:0x000000000066a8>`) however, i think that *will* work once i move the helper into the ApplicationController as you suggested below.

Comment (Sgtpluck) [reply app/controllers/application_controller.rb:2]: `current_user` and `user_signed_in?` are used in the view, so i am going to leave those in the helper

Comment (Sgtpluck) [reply app/controllers/application_controller.rb:2]: oh actually, since i use bypass_auth in the `current_user` method i don't think i can hm

Comment (Sgtpluck) [reply app/controllers/application_controller.rb:2]: oh wait duh you answered that below, ignore the last comment.

**Pull request merged**